core/src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -1901,14 +1901,68 @@ sepBy
 -------------------------------------------------------------------------------
 --
 -- | See 'Streamly.Internal.Data.Parser.sequence'.
---
--- /Unimplemented/
---
 {-# INLINE sequence #-}
-sequence ::
-    -- Foldable t =>
-    Fold m b c -> t (Parser m a b) -> Parser m a c
-sequence _f _p = undefined
+sequence :: MonadThrow m => Fold m b c -> D.Stream m (Parser m a b) -> Parser m a c
+sequence (Fold fstep finitial fextract) (D.Stream sstep sstate) =
+    Parser step initial extract
+    where
+    initial = do
+        fres <- finitial
+        case fres of
+            FL.Partial fs -> return $ IPartial (Nothing', sstate, fs)
+            FL.Done c -> return $ IDone c
+
+    -- state does not contain any parser
+    -- yield a new parser from the stream
+    step (Nothing', ss, fs) _ = do
+        sres <- sstep defState ss
+        case sres of
+            D.Yield p ss' -> return $ Continue 1 (Just' $ Left' p, ss', fs)
+            D.Stop -> do
+                c <- fextract fs
+                return $ Done 1 c
+            D.Skip ss' -> return $ Continue 1 (Nothing', ss', fs)
+
+    -- state contains a parser that has not yet been initialized
+    step (Just' (Left' (Parser pstep pinit pextr)), ss, fs) _ = do
```

Comment:
Why is `Left` required? This state is superseded by the one below, `Right`.

Reply:
Aah yes that is true. I somehow got stuck on logically separating the initial parser and the modified parser state. I think this whole portion will be overhauled, when using the separate state for the parser that you mentioned in a previous comment.

Reply:
I've removed the extra Either constructor now, since Just and nothing is enough.